### ISSUE TITLE
Preset default elliptic curve and accept handshake when client doesn't propose supported_groups

### ISSUE THREAD

**Issue by AmirHmZz** [openssl/openssl#22732]
According to RFC :

> A client that proposes ECC cipher suites may choose not to include these extensions.  In this case, the server is free to choose any one of the elliptic curves or point formats listed in [Section 5](https://datatracker.ietf.org/doc/html/rfc8422#section-5).

Also, as @mattcaswell mentions in #22703:

> This allows OpenSSL to just pick a curve if the client hasn't provided a list to choose from. There aren't any MUSTs here, so I think OpenSSL is also conformant in deciding to fail the handshake in this case. But it would be possible for us to add the ability to choose an arbitrary one and that would also be ok.

I think this feature could be very helpful specially for embedded devices which have minimal implementation of TLS protocol. In my opinion It would be better that openssl let developer to choose which curve to be used in order to maximize combability with different environment and devices.

**Comment by davidben** [comment]
There are serious compatibility risks to making guesses about the elliptic curve to use. If the server does not correctly divine the client's preferences, the connection will fail, when perhaps a non-ECC cipher suite would have worked. This would have to be an off-by-default option with lots of warnings on it.

Can you elaborate on why these embedded devices are failing to send the `supported_groups` extension? It doesn't make much sense to do this as a "minimal" implementation because the embedded device, it sounds like, is still implementing ECC. It's just refusing to send a few extra bytes in the ClientHello to tell you what it's doing. Embedded devices should be perfectly capable of sending this. It can even just be a hardcoded string.

The contortions in the specification around these fields being optional is mostly a legacy thing in TLS, when people were worried about being unable to send extensions for compatibility reasons. These concerns are long, long gone and TLS 1.3 removes all this. (There were also some negative consequences to this, e.g. those allowances in the signature algorithms extension led directly to some old OpenSSL bugs in signature algorithm negotiation going unnoticed for a while.)

I'm actually surprised we didn't fix this in RFC 8422 when we updated RFC 4492. An oversight, alas. Had I noticed this at the time, I would have gotten that fixed. :-)

**Comment by AmirHmZz** [comment]
> There are serious compatibility risks to making guesses about the elliptic curve to use. If the server does not correctly divine the client's preferences, the connection will fail, when perhaps a non-ECC cipher suite would have worked. This would have to be an off-by-default option with lots of warnings on it.

Yes, you are totally right. In my case, I know which GSM module is going to connect to the server with theses specific default. So, I know which curves those specific modules support and I'm sure those modules support that preset elliptic curve. I mean it is not going to be a general purpose feature to be used for all kind of clients. The developer **MUST** know why he wants to use this feature with knowledge of the clients' capabilities.


> Can you elaborate on why these embedded devices are failing to send the supported_groups extension? It doesn't make much sense to do this as a "minimal" implementation because the embedded device, it sounds like, is still implementing ECC. It's just refusing to send a few extra bytes in the ClientHello to tell you what it's doing. Embedded devices should be perfectly capable of sending this. It can even just be a hardcoded string.

You are completely right. Sending `supported_groups` should not be something difficult to for embedded devices if they are supporting ECC. Unfortunately, GSM modules' manufacturers doesn't provide the firmware's source code or even their SDK files. The only thing which is public is the compiled firmware. We've communicated with manufacturers and most of them don't have any plan to provide update for this reason and the rest of them just told us they will be working on that. So, we cannot fix these small problems on client side and all we hope to do is on the server side.


> The contortions in the specification around these fields being optional is mostly a legacy thing in TLS, when people were worried about being unable to send extensions for compatibility reasons. These concerns are long, long gone and TLS 1.3 removes all this.

For TLS1.2, RFC seems to allow client not to send `supported_groups`. So I think It is not much bad to have that feature for TLS1.2 only. If TLS1.3 removes these allowances even newer modules with TLS1.3 support have to send `supported_groups` and everything is fine then, too.

Another question had just popped into my mind: Is there any version/commit of OpenSSL that picks a default elliptic curve if client doesn't provide one?

**Comment by mattcaswell** [comment]
> Another question had just popped into my mind: Is there any version/commit of OpenSSL that picks a default elliptic curve if client doesn't provide one?

No.

**Comment by kasteelharry** [comment]
I have encountered a similar issue where the manufacturer does not send the supported groups extension with TLS 1.2. Were you able to find a solution for your issue?

**Comment by davidben** [comment]
Have you contacted the manufacturer to start sending the extension? Since any such option in OpenSSL can never be enabled by default, it is far better for the clients to behave in a standard way. Of course, you may be stuck with existing clients for a while, but we should make sure this is fixed for clients going forward.

**Comment by kasteelharry** [comment]
> Have you contacted the manufacturer to start sending the extension? Since any such option in OpenSSL can never be enabled by default, it is far better for the clients to behave in a standard way. Of course, you may be stuck with existing clients for a while, but we should make sure this is fixed for clients going forward.

Unfortunately the client cannot (or won't) send the extension. I agree that it is far better to have them send the extension, however, that is not up to us and we are stuck waiting until they do. Do you have any suggestions that might work in this case?

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
